### PR TITLE
POC e2e test

### PR DESCRIPTION
**Pull request by rutsky**
Based on PR #2134, which is based on #2105. Relevant changes: https://github.com/buildbot/buildbot/commit/df8900fb239afec7d924dff008a5d020e6749130

POC of e2e tests. Creates master, creates worker, starts master, starts worker, shutdowns worker, shutdowns master.

---

Current progress:

* [x] Setup/teardown tests
* [x] Run tests in daemon and non-daemon mode (Buildbot/Twisted doesn't handle daemonization Windows)
* [x] Tests with deprecated Buildslave
* [x] Tests with pure Python 3 worker
* [ ] Tests with pure Python 2 worker
* [ ] Tests with single-file worker
* [ ] Fix tests on Windows

### CODE REVIEW

File: master/buildbot/test/e2e/test_master_worker_create_start_stop.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        yield self.tearDownDirs()
+
+    def test_master_worker_setup(self):
+        os.chdir(self.projectdir)
+
+        master_dir = "master-dir"
+        worker_dir = "worker-dir"
+
+        # Create master.
+        stdout = subprocess.check_output(
```
Comment (tardyp) [inline master/buildbot/test/e2e/test_master_worker_create_start_stop.py]: check_output does not exist in python 2.6

Comment (rutsky) [inline master/buildbot/test/e2e/test_master_worker_create_start_stop.py]: I'll add [subprocess32](https://pypi.python.org/pypi/subprocess32/) to master test dependencies then, and use it on Python 2.

File: master/buildbot/test/e2e/test_master_worker_create_start_stop.py [hunk 2]
```diff
@@ -0,0 +1,921 @@
...
+from __future__ import print_function
+
+import os
+import re
+import socket
+import textwrap
+
+from fastjsonrpc.client import Proxy
+from fastjsonrpc.client import jsonrpc
+from twisted.internet import reactor as global_reactor
```
Comment (sa2ajj) [inline master/buildbot/test/e2e/test_master_worker_create_start_stop.py]: missing empty line

File: master/buildbot/test/e2e/test_master_worker_create_start_stop.py [hunk 3]
```diff
@@ -0,0 +1,921 @@
...
+from twisted.internet import defer
+from twisted.internet import task
+from twisted.internet import utils
+from twisted.python import log
+from twisted.trial import unittest
+from txrequests import Session
```
Comment (sa2ajj) [inline master/buildbot/test/e2e/test_master_worker_create_start_stop.py]: missing empty line

File: master/buildbot/test/util/decorators.py [hunk 4]
```diff
@@ -61,3 +61,18 @@ def closure(test):
...
     return closure
+
+
+def skipUnlessInstalled(module_name, error_message=None):
+    if error_message is None:
+        error_message = (
+            "module '{0}' is not available, "
+            "install appropriate package").format(module_name)
+
+    def closure(test):
```
Comment (sa2ajj) [inline master/buildbot/test/util/decorators.py]: i'd use `functools.wraps` here.

File: master/setup.py [hunk 5]
```diff
@@ -450,6 +450,8 @@ def define_plugin_entries(groups):
         'txgithub',
         'ramlfications',
         'mock',
+        'shutilwhich',
```
Comment (sa2ajj) [inline master/setup.py]: there seem no condition on the Python version here (unlike the place where it's used)

File: master/buildbot/test/e2e/test_master_worker_create_start_stop.py [hunk 6]
```diff
@@ -0,0 +1,921 @@
...
+
+from buildbot.test.util import dirs
+from buildbot.test.util.decorators import skipUnlessInstalled
+from buildbot.test.util.decorators import skipUnlessPlatformIs
+
+try:
+    from shutil import which
```
Comment (sa2ajj) [inline master/buildbot/test/e2e/test_master_worker_create_start_stop.py]: this is a standard library module, so the whole thing should go *before* 3rd party modules.

File: master/buildbot/test/e2e/test_master_worker_create_start_stop.py [hunk 7]
```diff
@@ -0,0 +1,921 @@
...
+except ImportError:
+    # Backport of shutil.which() from Python 3.3.
+    from shutilwhich import which
+
+try:
+    from textwrap import indent
```
Comment (sa2ajj) [inline master/buildbot/test/e2e/test_master_worker_create_start_stop.py]: ditto

File: master/buildbot/test/e2e/test_master_worker_create_start_stop.py [hunk 8]
```diff
@@ -0,0 +1,921 @@
...
+            caller.stop()
+
+            raise RuntimeError("Timeout")
+
+    start_time = reactor.seconds()
+    caller = task.LoopingCall(next_try)
+    return caller.start(check_interval, now=True)
+
+
+def _read_dir_contents(dir):
```
Comment (sa2ajj) [inline master/buildbot/test/e2e/test_master_worker_create_start_stop.py]: `dir` is a standard function...  why not use `dirname`?

File: master/buildbot/test/e2e/test_master_worker_create_start_stop.py [hunk 9]
```diff
@@ -0,0 +1,921 @@
...
+    contents = {}
+    for root, _, files in os.walk(dir):
+        for name in files:
+            fn = os.path.join(root, name)
```
Comment (sa2ajj) [inline master/buildbot/test/e2e/test_master_worker_create_start_stop.py]: fn -> filename

(I'm curious what, for example, `pylint` has to say about this file...)

File: master/buildbot/test/e2e/test_master_worker_create_start_stop.py [hunk 10]
```diff
@@ -0,0 +1,921 @@
...
+            # Logs can be stored on AppVeyor and CircleCI, we can move
+            # e2e tests there if we don't want such output.
+            print("Test failed, output:")
+            print("-" * 80)
+            print("\n".join(self.logs))
+            print("-" * 80)
+            self._print_all_logs()
+
+            os.chdir(self._origcwd)
+
```
Comment (sa2ajj) [inline master/buildbot/test/e2e/test_master_worker_create_start_stop.py]: please remove empty line

File: master/buildbot/test/e2e/test_master_worker_create_start_stop.py [hunk 11]
```diff
@@ -0,0 +1,921 @@
...
+        if self.master_dir is not None:
+            try:
+                yield self._run_command(['buildbot', 'stop', self.master_dir])
+            except Exception:
+                # Ignore errors.
+                pass
+
+    def _print_all_logs(self):
+        """Print all started services Twistd logs"""
+
```
Comment (sa2ajj) [inline master/buildbot/test/e2e/test_master_worker_create_start_stop.py]: ditto

File: master/buildbot/test/e2e/test_master_worker_create_start_stop.py [hunk 12]
```diff
@@ -0,0 +1,921 @@
...
+            f.write(master_cfg)
+
+    @defer.inlineCallbacks
+    @skipUnlessInstalled('buildbot_worker',
+                         "buildbot-worker package is not installed")
+    def test_master_worker_setup(self):
+        """Create master and worker (with default pyflakes configuration),
+        start them, stop them.
+        """
+
```
Comment (sa2ajj) [inline master/buildbot/test/e2e/test_master_worker_create_start_stop.py]: ditto

File: master/buildbot/test/e2e/test_master_worker_create_start_stop.py [hunk 13]
```diff
@@ -0,0 +1,921 @@
...
+            [^\n]+ Got workerinfo from 'example-worker'
+            """)
+
+        self.assertTrue(
+            re.search(worker_connection_re, log, re.MULTILINE),
+            msg="Log doesn't match:\n{0}\nLog:\n{1}".format(
+                indent(worker_connection_re, "    "),
+                indent(log, "    ")))
+
+        self.success = True
```
Comment (sa2ajj) [inline master/buildbot/test/e2e/test_master_worker_create_start_stop.py]: I am not sure what this assignment is for.  Could you please elaborate?

**Comment by rutsky** [comment]
Updated. Let's see how it test will go. Still WIP.

**Comment by tardyp** [comment]
I would prefer we continue on the smokes directory which also creates
master and worker. But also test via the UI

Le lun. 23 janv. 2017 22:43, Vladimir Rutsky <notifications@github.com> a
écrit :

**Comment by rutsky** [comment]
@tardyp does smokes runs on Windows?

I used this E2E tests to test that `buildbot-master` works with `buildbot-worker` and `buildbot-slave` on Linux and Windows.

Currently I will need it (or any other e2e tests) to test that single-file distribution of master and worker works on Linux and Windows.

File: master/docs/relnotes/index.rst [hunk 14]
```diff
@@ -1,117 +1,132 @@
-Release Notes
-~~~~~~~~~~~~~
+Release Notes for Buildbot ``|version|``
```
Comment (tardyp) [inline master/docs/relnotes/index.rst]: looks like the release notes parts are not wanted

Comment (rutsky) [inline master/docs/relnotes/index.rst]: Yes, this PR needs lots of rework since I initially made it half a year ago. I will work on this.

**Comment by rutsky** [comment]
@tardyp I think running Selenium tests out of this tests should work and will be a good fit, but I need to take a closer look at current smokes implementation.

**Comment by tardyp** [comment]
@rutsky see the ./smokes/run.sh

The actual selenium tests expect a master to run on localhost:8010 with an attached worker.
They can be run using the following commands inside the smokes directory

yarn install
./node_modules/protractor/bin/webdriver-manager update
./node_modules/protractor/bin/protractor protractor.conf.js

The rest of the run.sh is quick and dirty install of a master and worker that shall be replaced by proper automation like you did.

I run them before the releases to make sure the UI is usable, and we can run a force build, rebuild, change triggered build. David is working on more tests.
I've not yet implemented it in buildbot travis, but that is the plan.

File: master/buildbot/test/util/decorators.py (hunk 4 above)
Comment (rutsky) [inline master/buildbot/test/util/decorators.py]: `closure` is decorator, there is no need for `functools.wraps`.

File: master/buildbot/test/e2e/test_master_worker_create_start_stop.py (hunk 2 above)
Comment (rutsky) [inline master/buildbot/test/e2e/test_master_worker_create_start_stop.py]: @sa2ajj why there should be empty line?

File: master/buildbot/test/e2e/test_master_worker_create_start_stop.py (hunk 3 above)
Comment (rutsky) [inline master/buildbot/test/e2e/test_master_worker_create_start_stop.py]: @sa2ajj there is already single empty line here

File: master/buildbot/test/util/decorators.py (hunk 4 above)
Comment (tardyp) [inline master/buildbot/test/util/decorators.py]: right, as it does not replace the function

File: master/buildbot/test/e2e/test_master_worker_create_start_stop.py [hunk 15]
```diff
@@ -0,0 +1,921 @@
...
+c['builders'] = [
+    util.BuilderConfig(name='runtests', workernames=['example-worker'],
+        factory=factory),
+]
+
+c['buildbotURL'] = 'http://localhost:8010/'
+c['www'] = dict(port=8010, plugins={})
+c['db'] = {
+    'db_url' : 'sqlite:///state.sqlite',
+}
```
Comment (tardyp) [inline master/buildbot/test/e2e/test_master_worker_create_start_stop.py]: Please dont forget to disable BuildbotNetUsageData on those tests. I can see a spike of 150 installs on events.buildbot.net :)

Comment (rutsky) [inline master/buildbot/test/e2e/test_master_worker_create_start_stop.py]: Good point, I will fix this, thanks! :)

Comment (rutsky) [inline master/buildbot/test/e2e/test_master_worker_create_start_stop.py]: Usage reporting should be disabled now.

**Pull request closed without merging**